### PR TITLE
[Merged by Bors] - refactor: generalize instances of the continuous functional calculus

### PR DESCRIPTION
**Pull request by j-loreaux**
This refactors instances of the continuous functional calculus slightly. In particular, the derived instances for `IsSelfAdjoint` and `0 ≤ ·` will now trigger automatically once the type class conditions are satisfied, and these are generic, not specific to C⋆-algebras. For nonnegative elements, this utilizes the new type class `NonnegSpectrumClass`.

In addition, we show that the restricted continuous functional calculi are equal to their non-restricted counterparts. 

---

- [x] depends on: #11957 
- [x] depends on: #11958

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/NormedSpace/Star/ContinuousFunctionalCalculus/Instances.lean [hunk 1]
```diff
@@ -245,43 +282,74 @@ lemma spectrum_star_mul_self_nonneg {b : A} : ∀ x ∈ spectrum ℝ (star b * b
...
+  [NonnegSpectrumClass ℝ A]
+
+lemma cfcHom_nnreal_eq_restrict {a : A} (ha : 0 ≤ a) :
+    cfcHom ha = (SpectrumRestricts.nnreal_of_nonneg ha).starAlgHom
+      (cfcHom (IsSelfAdjoint.of_nonneg ha)) := by
+  have := UniqueContinuousFunctionalCalculus.compactSpace_spectrum a (R := ℝ)
+  apply (SpectrumRestricts.nnreal_of_nonneg ha).cfcHom_eq_restrict isometry_subtype_coe
+
+lemma cfc_nnreal_eq_real {a : A} (f : ℝ≥0 → ℝ≥0) (ha : 0 ≤ a := by cfc_tac)  :
+    cfc f a = cfc (fun x ↦ f x.toNNReal : ℝ → ℝ) a := by
```
Comment (dupuisf) [inline Mathlib/Analysis/NormedSpace/Star/ContinuousFunctionalCalculus/Instances.lean:347]: Do we also want `cfc_nnreal_eq_complex` for completeness?

**Review by dupuisf — COMMENTED** [review]
Looks good!

bors d+

File: Mathlib/Analysis/NormedSpace/Star/ContinuousFunctionalCalculus/Instances.lean (hunk 1 above)
Comment (j-loreaux) [reply Mathlib/Analysis/NormedSpace/Star/ContinuousFunctionalCalculus/Instances.lean:347]: I think not? There is no direct coercion from NNReal to Complex anyway. As for the other direction, you could use nnnorm, but then you make it harder to stop midway in Real. If we use it a lot and it's a pain to go through both lemmas, then we can add a direct one.

**Pull request closed without merging**